### PR TITLE
El Boppo the Clown punching bag game

### PR DESCRIPTION
**Pull request by thoys**
El Boppo the Clown is a game in which you have a punching bag in the shape of a clown. The game comes with a timer, scoreboard and sound effects.

In this PR a feature has been added to the teleport.js script:
  -  you can now ignore an entity in the teleport.js script, which allows you to teleport while having an entity attached that covers your hands in such a way it would normally block the teleport laser. 
     - To add to the ignore list:
        ```Messages.sendLocalMessage('Hifi-Teleport-Ignore-Add', 'YourEntityIDHere');```
     - To remove from the ignore list:
        ```Messages.sendLocalMessage('Hifi-Teleport-Ignore-Remove', 'YourEntityIDHere');```
## QA test plan
 For this test you need an HMD and hand-controllers:
 - Use the build of this PR ( because changes to teleport.js have been made )
 - Run the following script to create the El Boppo game: ```https://rawgit.com/thoys/hifi/bdabc21d57a65a218961fa42c050da147c79b3e2/unpublishedScripts/marketplace/boppo/createElBoppo.js```
 - While entering the zone and HMD is enabled, the boxing gloves will auto-attach to your hands.
 - Teleporting should still work while the boxing gloves are being worn.
 - To start, give the punching bag an initial punch, you should hear a bell and the game music should start now, as well as the timer will start running.
 - Every punch should add a point to the score.
 - In the last 3 seconds you should hear a ticking sound. (each second)
 - When the game ends you should hear the end bell, and the music should stop.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: febe25511b12fc1830448a3cca1790653bdf7a86 available at: https://gist.github.com/bd8191a5d710b71ee216b13e7368f3c9<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9937-febe255.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9937-febe255.dmg

**Comment by thoys** [comment]
@hifi-gustavo , test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: febe25511b12fc1830448a3cca1790653bdf7a86 available at: https://gist.github.com/bd8191a5d710b71ee216b13e7368f3c9<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9937-febe255.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9937-febe255.dmg

**Comment by claudioshg** [comment]
Testing

**Comment by claudioshg** [comment]
### Desktop PC - Windows 10 Home - CPU 4.0GHz Intel i7 - GTX1070 - RAM 16GB - Tested with Oculus Rift

Tested: Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9937-febe255.exe

I loaded the script and walked into the zone. No gloves appeared on my hands. The game is not activating.

File: unpublishedScripts/marketplace/boppo/boppoServer.js [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+                    _this.hit();
+                } else if (message === MESSAGE_ENTER_ZONE && !_isGameRunning) {
+                    _this.resetBoppo();
+                } else if (message === MESSAGE_UNLOAD_FIX && _isInitialized) {
+                    _this.unload();
+                }
+            }
+        };
+
+        BoppoServer = function () {
```
Comment (huffman) [inline unpublishedScripts/marketplace/boppo/boppoServer.js]: Another ctor inside a ctor? Is there a benefit to this pattern?

File: unpublishedScripts/marketplace/boppo/lookAtEntity.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
+//
+//  lookAtTarget.js
```
Comment (ZappoMan) [inline unpublishedScripts/marketplace/boppo/lookAtEntity.js]: is there a reason you included this script in this PR?

Comment (thoys) [inline unpublishedScripts/marketplace/boppo/lookAtEntity.js]: Yes, it is used to make Boppo face forward

Comment (thoys) [inline unpublishedScripts/marketplace/boppo/lookAtEntity.js]: It is included in the entity script

File: unpublishedScripts/marketplace/boppo/createElBoppo.js [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+//  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+//
+
+/* globals SCRIPT_IMPORT_PROPERTIES */
+
+var MODELS_PATH = 'http://hifi-content.s3.amazonaws.com/DomainContent/Welcome%20Area/production/models/boxingRing/';
+var WANT_CLEANUP_ON_SCRIPT_ENDING = false;
+
+var getScriptPath = function(localPath) {
+    if (this.isCleanupAndSpawnScript) {
```
Comment (huffman) [inline unpublishedScripts/marketplace/boppo/createElBoppo.js]: where is this defined?

File: unpublishedScripts/marketplace/boppo/lookAtEntity.js (hunk 2 above)
Comment (ZappoMan) [inline unpublishedScripts/marketplace/boppo/lookAtEntity.js]: ok

**Comment by hifi-gustavo** [comment]
Build console output for commit: bdabc21d57a65a218961fa42c050da147c79b3e2 available at: https://gist.github.com/bd8191a5d710b71ee216b13e7368f3c9<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9937-bdabc21.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9937-bdabc21.dmg

File: scripts/system/controllers/teleport.js [hunk 4]
```diff
@@ -239,11 +240,11 @@ function Teleporter() {
         //    We might hit an invisible entity that is not a seat, so we need to do a second pass.
         //  * In the second pass we pick against visible entities only.
         //
-        var intersection = Entities.findRayIntersection(pickRay, true, [], [this.targetEntity], false, true);
+        var intersection = Entities.findRayIntersection(pickRay, true, ignoredEntities, [this.targetEntity], false, true);
```
Comment (thoys) [inline scripts/system/controllers/teleport.js]: Hehe, yeh let me fix that.

File: unpublishedScripts/marketplace/boppo/boppoClownEntity.js [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+(function() {
+    var SFX_PREFIX = 'http://hifi-content.s3-us-west-1.amazonaws.com/caitlyn/production/elBoppo/sfx/';
+    var CHANNEL_PREFIX = 'io.highfidelity.boppo_server_';
+    var PUNCH_SOUNDS = [
+        'punch_1.wav',
+        'punch_2.wav'
+    ];
+    var PUNCH_COOLDOWN = 300;
+
+    Script.include(Script.resolvePath('lookAtEntity.js?v2'));
```
Comment (thoys) [inline unpublishedScripts/marketplace/boppo/boppoClownEntity.js]: Sure, let me remove those

**Comment by hifi-gustavo** [comment]
Build console output for commit: 535d4f5deaa8fa24ca43f03e85536641004018e1 available at: https://gist.github.com/bd8191a5d710b71ee216b13e7368f3c9<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9937-535d4f5.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9937-535d4f5.dmg

File: unpublishedScripts/marketplace/boppo/boppoClownEntity.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        var getOwnBoppoUserData = function() {
+            try {
+                return JSON.parse(Entities.getEntityProperties(_entityID, ['userData']).userData).Boppo;
+            } catch (e) {
+                // e
+            }
+            return {};
+        };
+
+        BoppoClownEntity = function () {
```
Comment (thoys) [inline unpublishedScripts/marketplace/boppo/boppoClownEntity.js]: Yes, trick to make it work more like a class self sustaining class, gave me no linting errors. Personally I like the pattern. It feels more encapsulated and the `_` variables are better protected than the other way that I used to do it like. Let me know if it bothers you or is offensive in some way :)

File: unpublishedScripts/marketplace/boppo/boppoServer.js (hunk 1 above)
Comment (thoys) [inline unpublishedScripts/marketplace/boppo/boppoServer.js]: see above

**Comment by claudioshg** [comment]
### Desktop PC - Windows 10 Home - CPU 4.0GHz Intel i7 - GTX1070 - RAM 16GB - Tested with Oculus Rift

Tested: Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9937-febe255.exe

- PASS: While entering the zone and HMD is enabled, the boxing gloves will auto-attach to your hands.
- _FAIL: Teleporting should still work while the boxing gloves are being worn._
- PASS: To start, give the punching bag an initial punch, you should hear a bell and the game music should start now, as well as the timer will start running.
- PASS: Every punch should add a point to the score.
- PASS: In the last 3 seconds you should hear a ticking sound. (each second)
- PASS: When the game ends you should hear the end bell, and the music should stop.

2 FAILS:
- 1) Teleporting doesn't work with the gloves on, because it looks like the gloves are blocking the laser pointer.
- 2) When I back away from the zone, and the gloves come off, teleporting does not work.
The laser points, but the teleportation function is off.
cpr9937

NOTES:
- When I push the hold buttons on my hand device, (the middle finger buttons) the gloves drop to the floor. Don't know if that is an intended function.
- The ticking sound, for the last 3 seconds, is very low comparred to all other sounds.
- Thanks for the workout!

**Comment by hifi-gustavo** [comment]
Build console output for commit: a8740614e23db29b5e6c7a67ea18e0e9c8533e6e available at: https://gist.github.com/bd8191a5d710b71ee216b13e7368f3c9<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9937-a874061.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9937-a874061.dmg

**Comment by claudioshg** [comment]
Tested: Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9937-a874061.exe

PASS
Teleportation works.
The game and everything else functions as it did before.

**Comment by hifi-gustavo** [comment]
Build console output for commit: b60534a85567cb61b32dfce5f9b52595586acd38 available at: https://gist.github.com/bd8191a5d710b71ee216b13e7368f3c9<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9937-b60534a.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9937-b60534a.dmg

**Comment by RyanDowne** [comment]
@huffman is this dev approved?

**Comment by huffman** [comment]
@RyanDowne There were a couple changes that @thoys and I had talked about on Slack. I will ping him directly to see where we are at on that

**Comment by hifi-gustavo** [comment]
Build console output for commit: dff67886d099d66f29cfc146651c66810466b092 available at: https://gist.github.com/bd8191a5d710b71ee216b13e7368f3c9<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9937-dff6788.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9937-dff6788.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: dff67886d099d66f29cfc146651c66810466b092 available at: https://gist.github.com/bd8191a5d710b71ee216b13e7368f3c9<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9937-dff6788.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9937-dff6788.dmg

**Pull request merged**